### PR TITLE
Feature/functional methods

### PR DESCRIPTION
**Pull request by pmosk**

### CODE REVIEW

File: src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.Concat.cs [hunk 1]
```diff
@@ -0,0 +1,58 @@
+namespace System;
+
+partial struct FlatArray<T>
+{
+    public FlatArray<T> Concat(FlatArray<T> other)
+    {
+        if (IsEmpty)
```
Comment (andreise) [inline src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.Concat.cs:7]: Consider rewriting the implementation in a way like this:

```cs
        if (other.length == default)
        {
            return this;
        }

        if (length == default)
        {
            return other;
        }

        var resultItems = new T[length + other.length];
        Array.Copy(items!, resultItems, length);
        Array.Copy(other.items!, 0, resultItems, length, other.length);

        return new(resultItems, default);
```

For the following reasons:
1. Using `Array.Copy`, the most efficient for the general cases. See a comment starting at [FlatArray.ToCollection.Explicit.cs#L61](https://github.com/pfpack/pfpack-core-flatcollections/blob/11f5f0a3ce3c46a2bf2a07df3eb952f72c892410/src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.ToCollection.Explicit.cs#L61) for the details.
2. The 'main' argument is `this`, `other` is the secondary. So, returning firstly `this` if the `other` empty looks more native.
3. There are lots of low-level work in the `FlatArray` implementation, and using the public properties there breaks the uniformity. So, currently there is an approach to use the same style low-level code everywhere in the implementation.

File: src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.Concat.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        {
+            resultItems[length + j] = other.items![j];
+        }
+
+        return new(resultItems, default);
+    }
+
+    public FlatArray<T> Concat(params T[] other)
+    {
+        if (IsEmpty)
```
Comment (andreise) [inline src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.Concat.cs:34]: Consider rewriting the implementation in a way like this:

```cs
        if (other is null || other.Length == default)
        {
            return this;
        }

        if (length == default)
        {
            return other;
        }

        var resultItems = new T[length + other.Length];
        Array.Copy(items!, resultItems, length);
        Array.Copy(other, 0, resultItems, length, other.Length);

        return new(resultItems, default);
```

For the following reasons:
1. Using `Array.Copy`, the most efficient for the general cases. See a comment starting at [FlatArray.ToCollection.Explicit.cs#L61](https://github.com/pfpack/pfpack-core-flatcollections/blob/11f5f0a3ce3c46a2bf2a07df3eb952f72c892410/src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.ToCollection.Explicit.cs#L61) for the details.
2. The 'main' argument is `this`, `other` is the secondary. So, returning firstly `this` if the `other` empty looks more native.
3. There are lots of low-level work in the `FlatArray` implementation, and using the public properties there breaks the uniformity. So, currently there is an approach to use the same style low-level code everywhere in the implementation.

File: src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.Concat.cs (hunk 1 above)
Comment (andreise) [reply src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.Concat.cs:7]: Also consider adding to the [InnerArrayHelper](https://github.com/pfpack/pfpack-core-flatcollections/blob/11f5f0a3ce3c46a2bf2a07df3eb952f72c892410/src/flatcollections-array/FlatArray/FlatArray.T/InnerArrayHelper/FlatArray.InnerArrayHelper.cs) a low-level method like this:

```cs
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        internal static T[] Concat(T[] array1, int length1, T[] array2, int length2)
        {
            Debug.Assert(length1 >= 0 && length1 <= array1.Length);
            Debug.Assert(length2 >= 0 && length2 <= array2.Length);

            var result = new T[length1 + length2];
            Array.Copy(array1, result, length1);
            Array.Copy(array2, 0, result, length1, length2);
            return result;
        }
```

So, the `Concat` implementation will be reduced to:

```cs
        if (other.length == default)
        {
            return this;
        }

        if (length == default)
        {
            return other;
        }

        var resultItems = InnerArrayHelper.Concat(items!, length, other.items!, other.length);
        return new(resultItems, default);
```

File: src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.Concat.cs (hunk 2 above)
Comment (andreise) [reply src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.Concat.cs:34]: Also consider adding to the [InnerArrayHelper](https://github.com/pfpack/pfpack-core-flatcollections/blob/11f5f0a3ce3c46a2bf2a07df3eb952f72c892410/src/flatcollections-array/FlatArray/FlatArray.T/InnerArrayHelper/FlatArray.InnerArrayHelper.cs) a low-level method like this:

```cs
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        internal static T[] Concat(T[] array1, int length1, T[] array2, int length2)
        {
            Debug.Assert(length1 >= 0 && length1 <= array1.Length);
            Debug.Assert(length2 >= 0 && length2 <= array2.Length);

            var result = new T[length1 + length2];
            Array.Copy(array1, result, length1);
            Array.Copy(array2, 0, result, length1, length2);
            return result;
        }
```

So, the `Concat` implementation will be reduced to:

```cs
        if (other is null || other.Length == default)
        {
            return this;
        }

        if (length == default)
        {
            return other;
        }

        var resultItems = InnerArrayHelper.Concat(items!, length, other, other.Length);
        return new(resultItems, default);
```

**Review by andreise — CHANGES_REQUESTED** [review]
Consider updating the `Concat` implementation in accordance with the review comments.

File: src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.Concat.cs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+        for (var j = 0; j < other.length; j++)
+        {
+            resultItems[length + j] = other.items![j];
+        }
+
+        return new(resultItems, default);
+    }
+
+    public FlatArray<T> Concat(params T[] other)
```
[the remark this answers is not in the document]
Comment (pmosk) [reply src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.Concat.cs]: Done

File: src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.Concat.cs (hunk 1 above)
Comment (pmosk) [reply src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.Concat.cs]: Implemented

File: src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.Concat.cs (hunk 2 above)
Comment (pmosk) [reply src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.Concat.cs]: Implemented

File: src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.Map.cs [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        {
+            return default;
+        }
+
+        var resultItems = new TResult[length];
+
+        for (int i = 0; i < length; i++)
+        {
+            resultItems[i] = map.Invoke(items![i]);
+        }
```
Comment (andreise) [inline src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.Map.cs:19]: Consider merging actual `main` into this one to make it replaceable by:

```suggestion
        InternalForEach((i, item) => resultItems[i] = map.Invoke(item));
```

File: src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.Map.cs [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        {
+            return default;
+        }
+
+        var resultItems = new TResult[length];
+
+        for (int i = 0; i < length; i++)
+        {
+            resultItems[i] = map.Invoke(items![i], i);
+        }
```
Comment (andreise) [inline src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.Map.cs:38]: Consider merging actual `main` into this one to make it replaceable by:

```suggestion
        InternalForEach((i, item) => resultItems[i] = map.Invoke(i, item));
```

**Review by andreise — CHANGES_REQUESTED** [review]
Consider updating `Filter`, `FlatMap`, and `Map` methods in accordance with the review comments.

File: src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.FlatMap.cs [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+            if (resultArray.IsEmpty)
+            {
+                continue;
+            }
+
+            for (var j = 0; j < resultArray.length; j++)
+            {
+                resultList.Add(resultArray.items![j]);
+            }
```
[the remark this answers is not in the document]
Comment (andreise) [reply src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.FlatMap.cs:30]: Also, if actual `main` if merged into this one, the cycle will be replaced but `InternalForEach`:

```cs
        InternalForEach(item =>
        {
            var resultArray = map.Invoke(item);

            if (resultArray.length == default)
            {
                return;
            }

            resultList.AddRange(new ArraySegment<TResult>(resultArray.items!, 0, resultArray.length));
        });
```

File: src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.Filter.cs [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        {
+            return default;
+        }
+
+        var startLength = length < InnerAllocHelper.DefaultPositiveCapacity ? length : InnerAllocHelper.DefaultPositiveCapacity;
+
+        var resultItems = new T[startLength];
+        var resultLength = 0;
+
+        for (int i = 0; i < length; i++)
```
Comment (andreise) [inline src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.Filter.cs:19]: If actual `main` is merged into this one, the cycle can be replaced by `InternalForEach`:
```cs
        InternalForEach(item =>
        {
            if (predicate.Invoke(item) is false)
            {
                return;
            }

            if (resultLength == resultItems.Length)
            {
                InnerBufferHelper.EnlargeBuffer(ref resultItems);
            }

            resultItems[resultLength++] = item;
        });
```

File: src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.Filter.cs [hunk 8]
```diff
@@ -0,0 +1,84 @@
+namespace System;
+
+partial struct FlatArray<T>
+{
+    public FlatArray<T> Filter(Predicate<T> predicate)
```
[the remark this answers is not in the document]
Comment (pmosk) [reply src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.Filter.cs]: Fixed

File: src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.Filter.cs [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+    {
+        _ = predicate ?? throw new ArgumentNullException(nameof(predicate));
+
+        if (length == default)
+        {
+            return default;
+        }
+
+        var startLength = length < InnerAllocHelper.DefaultPositiveCapacity ? length : InnerAllocHelper.DefaultPositiveCapacity;
```
[the remark this answers is not in the document]
Comment (pmosk) [reply src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.Filter.cs]: Fixed

File: src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.Filter.cs [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+    public FlatArray<T> Filter(Func<T, int, bool> predicate)
+    {
+        _ = predicate ?? throw new ArgumentNullException(nameof(predicate));
+
+        if (length == default)
+        {
+            return default;
+        }
+
+        var startLength = length < InnerAllocHelper.DefaultPositiveCapacity ? length : InnerAllocHelper.DefaultPositiveCapacity;
```
[the remark this answers is not in the document]
Comment (pmosk) [reply src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.Filter.cs]: Fixed

File: src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.Filter.cs [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+                continue;
+            }
+
+            if (resultLength == resultItems.Length)
+            {
+                InnerBufferHelper.EnlargeBuffer(ref resultItems);
+            }
+
+            resultItems[resultLength] = item;
+            resultLength++;
```
[the remark this answers is not in the document]
Comment (pmosk) [reply src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.Filter.cs]: Fixed

File: src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.Filter.cs [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
```
[the remark this answers is not in the document]
Comment (pmosk) [reply src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.Filter.cs]: Fixed

File: src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.FlatMap.cs [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+            return default;
+        }
+
+        var resultList = new List<TResult>();
+
+        for (int i = 0; i < length; i++)
+        {
+            var resultArray = map.Invoke(items![i]);
+
+            if (resultArray.IsEmpty)
```
[the remark this answers is not in the document]
Comment (pmosk) [reply src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.FlatMap.cs]: Fixed

File: src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.FlatMap.cs [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+            return default;
+        }
+
+        var resultList = new List<TResult>();
+
+        for (int i = 0; i < length; i++)
+        {
+            var resultArray = map.Invoke(items![i], i);
+
+            if (resultArray.IsEmpty)
```
[the remark this answers is not in the document]
Comment (pmosk) [reply src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.FlatMap.cs]: Fixed

File: src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.FlatMap.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (pmosk) [reply src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.FlatMap.cs]: Fixed

File: src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.FlatMap.cs [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+            {
+                continue;
+            }
+
+            for (var j = 0; j < resultArray.length; j++)
+            {
+                resultList.Add(resultArray.items![j]);
+            }
```
[the remark this answers is not in the document]
Comment (pmosk) [reply src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.FlatMap.cs]: Fixed

File: src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.Map.cs (hunk 4 above)
Comment (pmosk) [reply src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.Map.cs]: Fixed

File: src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.Map.cs (hunk 5 above)
Comment (pmosk) [reply src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.Map.cs]: Fixed

**Review by andreise — APPROVED** [review]

**Pull request merged**